File: src/main/ruby/truffleruby/core/truffle/debug.rb
# frozen_string_literal: true

# Copyright (c) 2017, 2023 Oracle and/or its affiliates. All rights reserved. This
# code is released under a tri EPL/GPL/LGPL license. You can use it,
# redistribute it and/or modify it under the terms of the:
#
# Eclipse Public License version 2.0, or
# GNU General Public License version 2, or
# GNU Lesser General Public License version 2.1.

module Truffle::Debug
  # Note: used by MSpec
  def self.show_backtraces
    $stderr.puts 'All Thread and Fiber backtraces:'
    Primitive.all_fibers_backtraces.each do |fiber, backtrace|
      $stderr.puts "#{fiber} of #{Primitive.fiber_thread(fiber)}", backtrace, nil
    end
  end

  class Breakpoint
    def initialize(handle)
      @handle = handle
    end

    def remove
      Truffle::Debug.remove_handle @handle
    end

  end

  def self.break(file, line, &block)
    Breakpoint.new(break_handle(File.expand_path(file), line, &block))
  end
end

File: src/main/c/cext/debug.c
/*
 * Copyright (c) 2021, 2023 Oracle and/or its affiliates. All rights reserved. This
 * code is released under a tri EPL/GPL/LGPL license. You can use it,
 * redistribute it and/or modify it under the terms of the:
 *
 * Eclipse Public License version 2.0, or
 * GNU General Public License version 2, or
 * GNU Lesser General Public License version 2.1.
 */
#include <truffleruby-impl.h>
#include <ruby/debug.h>


struct rb_debug_inspector_struct {
    VALUE backtrace;
    VALUE contexts; /* [[self, klass, binding], ...] */
    long backtrace_size;
};

VALUE rb_debug_inspector_backtrace_locations(const rb_debug_inspector_t *dc) {
  return dc->backtrace;
}

VALUE rb_debug_inspector_open(rb_debug_inspector_func_t func, void *data) {
  rb_debug_inspector_t dbg_context;
  VALUE backtrace = RUBY_CEXT_INVOKE("rb_thread_current_backtrace_locations");
  dbg_context.contexts = RUBY_CEXT_INVOKE("rb_debug_inspector_open_contexts");
  if (RARRAY_LENINT(backtrace) != RARRAY_LENINT(dbg_context.contexts)) {
    rb_raise(rb_eRuntimeError, "debug_inspector contexts and backtrace lengths are not equal");
  }
  dbg_context.backtrace = backtrace;
  dbg_context.backtrace_size = RARRAY_LEN(backtrace);
  return (*func)(&dbg_context, data);
}

static VALUE frame_get(const rb_debug_inspector_t *dc, long index) {
  if (index < 0 || index >= dc->backtrace_size) {
    rb_raise(rb_eArgError, "no such frame");
  }
  return rb_ary_entry(dc->contexts, index);
}

enum {
  CALLER_BINDING_SELF,
  CALLER_BINDING_CLASS,
  CALLER_BINDING_BINDING
};

VALUE rb_debug_inspector_frame_binding_get(const rb_debug_inspector_t *dc, long index) {
  VALUE frame = frame_get(dc, index);
  return rb_ary_entry(frame, CALLER_BINDING_BINDING);
}

VALUE rb_debug_inspector_frame_iseq_get(const rb_debug_inspector_t *dc, long index) {
  return Qnil;
}

VALUE rb_debug_inspector_frame_class_get(const rb_debug_inspector_t *dc, long index) {
  VALUE frame = frame_get(dc, index);
  return rb_ary_entry(frame, CALLER_BINDING_CLASS);
}

VALUE rb_debug_inspector_frame_self_get(const rb_debug_inspector_t *dc, long index) {
  VALUE frame = frame_get(dc, index);
  return rb_ary_entry(frame, CALLER_BINDING_SELF);
}
